From 18b74ce54afab45fcf2a7d4eb86bb2ce9db8cec8 Mon Sep 17 00:00:00 2001
From: Richard Henderson <rth@redhat.com>
Date: Tue, 21 Oct 2014 13:00:34 -0400
Subject: [PATCH 052/415] aarch64: Fix non-apple compilation

---
 src/aarch64/ffi.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/aarch64/ffi.c b/src/aarch64/ffi.c
index 5369ea4..cdb7816 100644
--- a/src/aarch64/ffi.c
+++ b/src/aarch64/ffi.c
@@ -782,7 +782,9 @@ ffi_prep_cif_machdep (ffi_cif *cif)
           }
     }
 
+#if defined (__APPLE__)
   cif->aarch64_nfixedargs = 0;
+#endif
 
   return FFI_OK;
 }
-- 
2.7.4.huawei.3

