From 9c12209d2eac40238eefb4255994277918e7eff1 Mon Sep 17 00:00:00 2001
From: Francis Ricci <francisjricci@gmail.com>
Date: Thu, 3 Aug 2017 10:46:28 -0700
Subject: [PATCH 252/411] Fix misaligned memory access in ffi_call_int

---
 src/x86/ffi64.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/x86/ffi64.c b/src/x86/ffi64.c
index 2603a3a..757930b 100644
--- a/src/x86/ffi64.c
+++ b/src/x86/ffi64.c
@@ -646,10 +646,10 @@ ffi_call_int (ffi_cif *cif, void (*fn)(void), void *rvalue,
 		  break;
 		case X86_64_SSE_CLASS:
 		case X86_64_SSEDF_CLASS:
-		  reg_args->sse[ssecount++].i64 = *(UINT64 *) a;
+		  memcpy (&reg_args->sse[ssecount++].i64, a, sizeof(UINT64));
 		  break;
 		case X86_64_SSESF_CLASS:
-		  reg_args->sse[ssecount++].i32 = *(UINT32 *) a;
+		  memcpy (&reg_args->sse[ssecount++].i32, a, sizeof(UINT32));
 		  break;
 		default:
 		  abort();
-- 
1.8.3.1

